test/categorical_algebra/CSetMorphisms.jl

```diff
@@ -37,6 +37,74 @@ add_edges!(h, [1,2], [2,1])
 @test force(compose(id(g), α)) == α
 @test force(compose(α, id(h))) == α
 
+# Limits
+#-------
+
+# Terminal object in Graph: the self-loop.
+term = Graph(1)
+add_edge!(term, 1, 1)
+@test ob(terminal(Graph)) == term
+
+# Products in Graph: unitality.
+lim = product(g, term)
+@test ob(lim) == g
+@test force(proj1(lim)) == force(id(g))
+@test force(proj2(lim)) ==
+  CSetTransformation((V=fill(1,4), E=fill(1,3)), g, term)
+
+# Product in Graph: two directed intervals (Reyes et al 2004, p. 48).
+I = Graph(2)
+add_edge!(I, 1, 2)
+prod = ob(product(I, I))
+@test nv(prod) == 4
+@test ne(prod) == 1
+@test src(prod) != tgt(prod)
+
+# Product in Graph: deleting edges by multiplying by an isolated vertex.
+g = Graph(4)
+add_edges!(g, [1,2,3], [2,3,4])
+g0 = ob(product(g, Graph(1)))
+@test nv(g0) == nv(g)
+@test ne(g0) == 0
+
+# Product in Graph: copying edges by multiplying by the double self-loop.
+cycle2 = Graph(1)
+add_edges!(cycle2, [1,1], [1,1])
+g2 = ob(product(g, cycle2))
+@test nv(g2) == nv(g)
+@test ne(g2) == 2*ne(g)
+@test src(g2) == repeat(src(g), 2)
+@test tgt(g2) == repeat(tgt(g), 2)
+
+# Equalizer in Graph from (Reyes et al 2004, p. 50).
+g, h = Graph(2), Graph(2)
+add_edges!(g, [1,2], [2,1])
+add_edges!(h, [1,2,2], [2,1,1])
+ϕ = CSetTransformation((V=[1,2], E=[1,2]), g, h)
+ψ = CSetTransformation((V=[1,2], E=[1,3]), g, h)
+@test is_natural(ϕ) && is_natural(ψ)
+eq = equalizer(ϕ, ψ)
+@test ob(eq) == I
+@test force(incl(eq)[:V]) == FinFunction([1,2])
+@test force(incl(eq)[:E]) == FinFunction([1], 2)
+
+# Pullback in Graph from (Reyes et al 2004, p. 53).
+# This test exposes an error in the text: using their notation, there should be
```

Comment:
Products and pullbacks of C-sets can be unintuitive compared to coproducts and pushouts. If the authors of this textbook had a program for computing limits, they would have caught this error!